src/plonk/permutation.rs

```diff
@@ -20,24 +18,49 @@ pub(crate) struct Argument {
 }
 
 impl Argument {
-    pub(crate) fn new(columns: Vec<Column<Any>>) -> Self {
-        Argument { columns }
+    pub(crate) fn new() -> Self {
+        Argument { columns: vec![] }
     }
 
     pub(crate) fn required_degree(&self) -> usize {
-        // The permutation argument will serve alongside the gates, so must be
-        // accounted for. There are constraints of degree 2 regardless of the
-        // number of columns involved. (It doesn't make sense to make a
-        // permutation argument with zero columns but to be rigorous we account
-        // for it here.)
-
         // degree 2:
         // l_0(X) * (1 - z(X)) = 0
         //
-        // degree columns + 1
-        // z(X) \prod (p(X) + \beta s_i(X) + \gamma)
-        // - z(omega^{-1} X) \prod (p(X) + \delta^i \beta X + \gamma)
-        std::cmp::max(self.columns.len() + 1, 2)
+        // We will fit as many polynomials p_i(X) as possible
+        // into the required degree of the circuit, so the
+        // following will not affect the required degree of
+        // this middleware.
+        //
+        // (1 - (l_last + l_blind)) * (
+        //   z(\omega X) \prod (p(X) + \beta s_i(X) + \gamma)
+        // - z(X) \prod (p(X) + \delta^i \beta X + \gamma)
+        // )
+        //
+        // On the first sets of columns, except the first
+        // column, we will do
+        //
+        // l_0(X) * (z(X) - z'(\omega^(last) X)) = 0
+        //
+        // where z'(X) is the permutation for the last set
+        // of columns.
+        //
+        // On the final set of columns, we will do
+        //
+        // degree 3:
+        // l_last(X) * (z'(X)^2 - z'(X)) = 0
+        //
+        // which will allow the last value to be zero to
+        // ensure the argument is perfectly complete.
+
+        // There are constraints of degree 3 regardless of the
+        // number of columns involved.
+        3
```

Comment:
I'm confused, I thought the idea was to pack as many columns per column-set as would fit in the degree bound.

Reply:
It is. However, if the number of columns is few enough that they contribute lower than degree 3, this sets the floor due to the `l_last` constraint.

What is confusing here is that this API previously returned the actually-used degree, whereas now it returns the "required" degree, and the actual degree will fluctuate based on the degree of the rest of the circuit. It would be helpful to document this API or rename it (to distinguish it from other `required_degree` APIs that behave differently).